<a href="https://colab.research.google.com/github/amadords/Projetos-Publicos/blob/master/Teste_de_Overfitting_no_classificador_Spotify.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Teste de Overfitting no classificador Spotify**
---


[![LinkedIn](https://img.shields.io/badge/LinkedIn-DanielSousaAmador-cyan.svg)](https://www.linkedin.com/in/daniel-sousa-amador)
[![GitHub](https://img.shields.io/badge/GitHub-amadords-darkblue.svg)](https://github.com/amadords)
[![Medium](https://img.shields.io/badge/Medium-DanielSousaAmador-white.svg)](https://daniel-s-amador.medium.com/)




Overfitting é, de uma forma bem direta, um **sobreajuste** do modelo aos dados, ou seja, o modelo de Machine Learning aprende a classificar tanto os modelos de treino que começa a **sofrer para generalizar** os dados de teste. [Aqui](https://bit.ly/30mKZ1q) você pode ver mais um exemplo de teste de overfitting.

O Overfitting é **muito comum** e normalmente a primeira forma de se conseguir avaliar é, testando com os dados de teste e com os próprios dados de treino.

Vale saber que overfitting é o exato **contraponto do underfitting**, que é o sobajuste, ou seja, quando o modelo fica fraco e não consegue nem explicar os dados de treino com alguma qualidade.

Não é dificil imaginar que é nesse meio termo que está o nosso modelo utilizável.

O overfitting é bem **mais comum que o underfitting**, mas vale o conhecimento sobre ambos.

Veja a imagem abaixo:

![over](https://experience.sap.com/wp-content/uploads/2018/02/illustrtaion-for-blog-06-700x300.jpg)


Não é, talvez, a melhor imagem para tratar do assunto, mas creio que dá para você compreender.

A figura **Fit** é aquela que consegue entender que há um padrão de dados e os consegue generalizar sem muita complexidade.

A **Undeffiting** não consegue separar os dados de nenhuma forma, é quase um algoritmo *lazy* ou *naive*, contudo a  culpa não é do algoritmo, mas da base de dados fraca.

**Overfitting** falar por si só, ele é muito complexo, por isso faz alguns "malabarismos" para classificar os dados.

* Aparentemente, esse modelo é o melhor para se utilizar, não?! Afinal, o que queremos é que um modelo acerte os dados. Mas vamos entender o porquê isso não é útil.
* Imagine que você tem dois grupos. Grupo **1.0** e grupo **2.0**; Se chegar um novo dado com um valor **1.0**, ele vai saber classificar como **1.0**, mas se chegar um valor **1.2**, por ele não saber generalizar, não vai saber onde colocar esse dados, por não ter aprendido a generalizar os dados, apenas trabalhar com os dados especificos.

## Checklist
1. Processos de Machine Learning
2. Teste de Overfitting (Você pode ir direto para esse ponto)

# 1. Processos de Machine Learning

**Importando a biblioteca pandas**

In [ ]:
import pandas as pd

**Carregando o dataset Spotify**

In [ ]:
dataset_spotify= pd.read_csv('https://raw.githubusercontent.com/amadords/data/main/data.csv', sep=',')

**Separando os dados**

In [ ]:
classes = dataset_spotify['target']
dataset_spotify.drop('target', axis=1, inplace=True)

**Criando função pra remover coluna sempre que necessário**

In [ ]:
def remove_features(lista_features):
    for i in lista_features:
        dataset_spotify.drop(i, axis=1, inplace=True)
    return 'OK'

**Visualizando os dados**

In [ ]:
dataset_spotify.head(3)

id  acousticness  danceability  ...  valence    song_title            artist
0   0        0.0102         0.833  ...    0.286      Mask Off            Future
1   1        0.1990         0.743  ...    0.588       Redbone  Childish Gambino
2   2        0.0344         0.838  ...    0.173  Xanny Family            Future

[3 rows x 16 columns]

**Removendo features não-representativas**

In [ ]:
remove_features(['id','song_title'])

'OK'

**Aplicando o get_dummies nos dados**

In [ ]:
dataset_spotify=pd.get_dummies(dataset_spotify, columns=['artist'], prefix=['artist'])

**Visualizando dados já com colunas removidas**

In [ ]:
dataset_spotify.head()

acousticness  danceability  ...  artist_카우칩스 The CowChips  artist_플랫핏 Flat Feet
0        0.0102         0.833  ...                         0                     0
1        0.1990         0.743  ...                         0                     0
2        0.0344         0.838  ...                         0                     0
3        0.6040         0.494  ...                         0                     0
4        0.1800         0.678  ...                         0                     0

[5 rows x 1356 columns]

**Dividindo os dados em treino e teste em 80/20**

In [ ]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(dataset_spotify,classes,test_size=0.2,random_state=123)

**Visualizando os dados de treino com as features**

In [ ]:
X_train

acousticness  ...  artist_플랫핏 Flat Feet
1505      0.004770  ...                     0
813       0.114000  ...                     0
615       0.030200  ...                     0
319       0.106000  ...                     0
320       0.021100  ...                     0
...            ...  ...                   ...
2012      0.001060  ...                     0
1346      0.000021  ...                     0
1406      0.503000  ...                     0
1389      0.705000  ...                     0
1534      0.623000  ...                     0

[1613 rows x 1356 columns]

**Visualizando os dados de treino com a classe**

In [ ]:
y_train

1505    0
813     1
615     1
319     1
320     1
       ..
2012    0
1346    0
1406    0
1389    0
1534    0
Name: target, Length: 1613, dtype: int64

**Visualizando os dados de teste com features**

In [ ]:
X_test

acousticness  ...  artist_플랫핏 Flat Feet
1423       0.00184  ...                     0
356        0.00855  ...                     0
854        0.93400  ...                     0
1445       0.09870  ...                     0
1647       0.00346  ...                     0
...            ...  ...                   ...
156        0.00212  ...                     0
883        0.19800  ...                     0
853        0.27100  ...                     0
903        0.02450  ...                     0
890        0.03280  ...                     0

[404 rows x 1356 columns]

**Visualizando os dados de teste com a classe**

In [ ]:
y_test

1423    0
356     1
854     1
1445    0
1647    0
       ..
156     1
883     1
853     1
903     1
890     1
Name: target, Length: 404, dtype: int64

**Treinando o classificador clf com SVM**

In [ ]:
# importação do algoritmo
from sklearn import svm
# instanciação do algoritmo
clf = svm.SVC().fit(dataset_spotify,classes)
# treinando o algoritmo
clf.fit(X_train,y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

# 2. Teste de Overfitting

**Importando bibliotecas**

In [ ]:
from sklearn.metrics import accuracy_score
from sklearn import metrics

**Predições de resultados com dados de teste**

In [ ]:
y_pred=clf.predict(X_test)

**Imprimindo acurácia com predições e dados de teste**

In [ ]:
metrics.accuracy_score(y_test,y_pred)

0.5668316831683168

**Predições de resultados com dados de treino**

In [ ]:
y_pred2=clf.predict(X_train)

**Imprimindo acurácia com predições e dados de treino**

In [ ]:
metrics.accuracy_score(y_train,y_pred2)

0.5523868567885927

### Houve Overfitting?

Nesse caso, não. Veja que os **resultados foram bem próximos** (56% e 55%). 

Se, ao alterar os parâmetros, os dados de treino começarem a melhorar o percentual da acurácia e os de teste continuarem estagnados ou não forem subindo na mesma, ou quase na mesma proporção, o modelo estará começando a sofrer overfitting.

O **Cientista de Dados** tem que observar e avaliar sempre!

# Obrigado!

Obrigado por ter disponibilizado um pouco do seu tempo e atenção aqui. Espero que, de alguma forma, tenha sido útil para seu crescimento. Se houver qualquer dúvida ou sugestão, não hesite em entrar em contato no [LinkedIn](https://www.linkedin.com/in/daniel-sousa-amador) e verificar meus outros projetos no [GitHub](https://github.com/amadords).


[![LinkedIn](https://img.shields.io/badge/LinkedIn-DanielSousaAmador-cyan.svg)](https://www.linkedin.com/in/daniel-sousa-amador)
[![GitHub](https://img.shields.io/badge/GitHub-amadords-darkblue.svg)](https://github.com/amadords)
[![Medium](https://img.shields.io/badge/Medium-DanielSousaAmador-white.svg)](https://daniel-s-amador.medium.com/)


<center><img width="90%" src="https://raw.githubusercontent.com/danielamador12/Portfolio/master/github.png"></center>